kafka/src/test/java/org/apache/metamodel/kafka/KafkaDataContextTest.java

```diff
@@ -0,0 +1,172 @@
+import org.apache.kafka.common.TopicPartition;
+import org.apache.metamodel.DataContext;
+import org.apache.metamodel.data.DataSet;
+import org.easymock.Capture;
+import org.easymock.EasyMock;
+import org.easymock.EasyMockSupport;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class KafkaDataContextTest extends EasyMockSupport {
+
+    @Test
+    public void testGetSchemaInfo() {
+        final ConsumerAndProducerFactory consumerFactory = createMock(ConsumerAndProducerFactory.class);
+
+        replayAll();
+
+        final Supplier<Collection<String>> topicSupplier = () -> Arrays.asList("foo", "bar");
+        final DataContext dc = new KafkaDataContext<>(String.class, String.class, consumerFactory, topicSupplier);
+
+        verifyAll();
+
+        Assert.assertEquals("[foo, bar]", dc.getDefaultSchema().getTableNames().toString());
+    }
+
+    @Test
+    public void testQueryWithoutOptimization() {
+        final ConsumerAndProducerFactory consumerFactory = createMock(ConsumerAndProducerFactory.class);
+        @SuppressWarnings("unchecked")
+        final Consumer<String, String> consumer = createMock(Consumer.class);
+
+        EasyMock.expect(consumerFactory.createConsumer("myTopic", String.class, String.class)).andReturn(consumer);
+
+        final List<PartitionInfo> partitionInfoList = new ArrayList<>();
+        partitionInfoList.add(new PartitionInfo("myTopic", 0, null, null, null));
+        partitionInfoList.add(new PartitionInfo("myTopic", 1, null, null, null));
+        partitionInfoList.add(new PartitionInfo("myTopic", 2, null, null, null));
+        partitionInfoList.add(new PartitionInfo("myTopic", 3, null, null, null));
+
+        EasyMock.expect(consumer.partitionsFor("myTopic")).andReturn(partitionInfoList);
+
+        final Capture<Collection<TopicPartition>> assignmentCapture = new Capture<>();
+        consumer.assign(EasyMock.capture(assignmentCapture));
+        consumer.seekToBeginning(EasyMock.anyObject());
+
+        final ConsumerRecords<String, String> consumerRecords1 = createConsumerRecords(1, 0, 10);
+        final ConsumerRecords<String, String> consumerRecords2 = createConsumerRecords(2, 20, 10);
+
+        EasyMock.expect(consumer.poll(1000)).andReturn(consumerRecords1);
+        EasyMock.expect(consumer.poll(1000)).andReturn(consumerRecords2);
+        EasyMock.expect(consumer.poll(1000)).andReturn(null);
+
+        consumer.unsubscribe();
+        consumer.close();
+
+        replayAll();
+
+        final Supplier<Collection<String>> topicSupplier = () -> Arrays.asList("myTopic");
+        final DataContext dc = new KafkaDataContext<>(String.class, String.class, consumerFactory, topicSupplier);
+
+        final DataSet dataSet = dc.query().from("myTopic").select("partition", "offset", "value").where("key").eq(
+                "key2").execute();
+        Assert.assertTrue(dataSet.next());
+        Assert.assertEquals("Row[values=[1, 2, value2]]", dataSet.getRow().toString());
```

Comment:
Please do not validate ResultSets via String serialization.
Without an explicit order by, the query processor should also be free to change ordering.

Reply:
Actually the order here is given by the mock Consumer object above in the code. It’s just an imitation/mock scenario of how the records may be delivered to the client. So in this case I think the asserted order also makes sense to include because we assert that they are all there and that they’re streamed out and not reordered.

Reply:
Correct me, if i am wrong: 
The apis the test should be programmed to not make any of those guarantees on both the serialization format and order.

When i modified the nested loop join i ran into a lot of those test. They failed as they assumed a specific order, so i rewrote working code in order to not change too many test.

In case of this particular test this is not a problem, as this is just a projection/selection, but in case of other operators, this can easily be  a problem, so i would avoid this kind of testing.

Reply:
You're not wrong :-) Kafka does give a few guarantees - the main one being that within the same partition, the offsets will be continuously incrementing. So a bit of ordering can be guaranteed, but obviously not across multiple partitions.

In this case I want to ensure that all my records are in deed there on the output. Do you think there is a nicer way that I can do that? I'm open for doing it differently, but this seems to me to be the most readable and easy way to do it.